### PR TITLE
John Splintercell: a gun that is only good for shooting out lights

### PR DESCRIPTION
**Pull request by Hatterhat**
## About The Pull Request
adds the SC/FISHER lightbreaker self-charging energy pistol, which does 0 damage. As a joke.
image

https://github.com/tgstation/tgstation/assets/31829017/97572baa-7421-4800-a60e-2db03f4adc6d

<details><summary>actual details, in case the video wasn't good enough:</summary>

unless you shoot at light fixtures,
image
floodlights,
image
or people with flashlights/seclites (even on helmets or guns) (it still does 0 damage, it just turns off their lights. not permanently)
image
it also works on crusher lights. and cyborg headlights. i don't have a screenshot for it.
doesn't work on flares though.

also it can shoot past machines and lockers
image
</details>
(also adds a variable for deciding how loud a dry fire sound is, in case you want to make your gun's empty sound be less loud.)

## Why It's Good For The Game

Adds a silly little tool for silly little men who either really hate lightbulbs or want to recreate the experience of being John Splintercell, the lightbulb-assasinating secret agent from Fork Echelon.

## Changelog

:cl:
add: The SC/FISHER disruptor pistol, a very compact, permanently silenced energy gun, is now stocked in Nanotrasen-accessible black markets with a price generally somewhere between 400 and 800 credits. Aspiring users are warned that it's really bad for trying to actually kill people. Caveat emptor.
add: Guns now have a dry_fire_sound_volume variable, allowing for guns to be less loud when trying to fire while empty.
fix: Closets and crates now properly count as structures for pass flags again.
/:cl:

### CODE REVIEW

**Comment by jlsnow301** [comment]
This could be made into an event or something. What if nightmare came in two flavors?

I don't think anyone wants to instantly valid themselves over the ability to shoot lightbulbs

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	pass_flags = PASSTABLE | PASSMACHINE | PASSSTRUCTURE
+	var/list/i_hate_lightbulbs = list(/obj/machinery/light,
+		/obj/machinery/power/floodlight,)
+
+/obj/projectile/energy/fisher/on_hit(atom/target, blocked, pierce_hit)
+	if(is_type_in_list(target, i_hate_lightbulbs))
+		damage = 25 // should guarantee light breaks
+		demolition_mod = 2 // should VERY guarantee light breaks
+	. = ..()
+	if(ishuman(target))
```
Comment (jlsnow301) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:17]: An early return here please

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		var/flickered = FALSE
+		for(var/obj/item/thingy in target.contents)
+			if(istype(thingy, /obj/item/flashlight))
+				var/obj/item/flashlight/light = thingy
+				if(light.on)
+					light.toggle_light()
+					flickered = TRUE
```
Comment (jlsnow301) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:24]: Add a continue here if the other conditions are not applicable to this path

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+			var/datum/component/seclite_attachable/has_a_light = thingy.GetComponent(/datum/component/seclite_attachable)
+			if(has_a_light && has_a_light.light.on)
```
Comment (jlsnow301) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:26]: Optional chain with ?

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/obj/projectile/energy/fisher
+	name = "attenuated kinetic force"
+	icon_state = null
+	damage = 0
+	damage_type = BRUTE
+	armor_flag = BOMB
+	range = 7
+	pass_flags = PASSTABLE | PASSMACHINE | PASSSTRUCTURE
+	var/list/i_hate_lightbulbs = list(/obj/machinery/light,
```
Comment (jlsnow301) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:9]: This must have a comment with ///

File: code/modules/projectiles/guns/energy/recharge.dm [hunk 5]
```diff
@@ -135,3 +135,23 @@
...
+	suppressed = TRUE
+	recharge_time = 1.2 SECONDS
+	ammo_type = list(/obj/item/ammo_casing/energy/fisher)
+
+/obj/item/gun/energy/recharge/fisher/examine_more(mob/user)
+	. = ..()
+	. += "The SC/FISHER is an illegally-modified kinetic accelerator cut down and refit into a disassembled miniature energy gun chassis, with its pressure chamber \
+	attenuated to launch kinetic bolts that resonate with glass in light fixtures and power connections in human-held flashlights - and <i>only</i> those. \
+	While some would argue that this is a really terrible design choice, others argue that it is very funny to be able to shoot at light sources. Caveat emptor."
+	return .
```
Comment (jlsnow301) [inline code/modules/projectiles/guns/energy/recharge.dm:157]:
```suggestion
```

This is implied

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/projectile/special/lightbreaker.dm (hunk 2 above)
Comment (Hatterhat) [reply code/modules/projectiles/projectile/special/lightbreaker.dm:24]: i'm assuming this means on the same level as the light toggling but if i'm wrong, uuuh, oops.

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		/obj/machinery/power/floodlight,)
+
+/obj/projectile/energy/fisher/on_hit(atom/target, blocked, pierce_hit)
+	if(is_type_in_list(target, i_hate_lightbulbs))
+		damage = 25 // should guarantee light breaks
+		demolition_mod = 2 // should VERY guarantee light breaks
+	. = ..()
+	if(!ishuman(target))
+		return
```
Comment (Ghommie) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:19]: remove.

Comment (Ghommie) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:19]: it shouldn't be resctricted to humans.

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	var/flickered = FALSE
+	for(var/obj/item/thingy in target.contents)
```
Comment (Ghommie) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:21]:
```suggestion
	if(!isliving(target))
		return
	var/mob/living/living_target = target
	for(var/obj/item/thingy as anything in living_target.get_equipped_items(include_pockets = TRUE, include_accessories = TRUE))
```
Something like this ought to be done instead.

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 8]
```diff
@@ -0,0 +1,33 @@
+/obj/projectile/energy/fisher
+	name = "attenuated kinetic force"
+	icon_state = null
```
Comment (Ghommie) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:3]: what about setting alpha to 0?

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Comment by Vincent983** [comment]
the black market idea sounds good, we already have a few other funny items there like the big-slappy

**Comment by DaydreamIQ** [comment]
If only ninjas could use guns, this'd be hilarious for them

**Comment by Serxule** [comment]
> the black market idea sounds good, we already have a few other funny items there like the big-slappy

this would be a good black market item, one of the harmless ones actually (considering you can get some strong (if clunky) items from there)

**Comment by Serxule** [comment]
also, is this a projectile, hitscan, or like a laser pointer?

**Comment by Hatterhat** [comment]
> also, is this a projectile, hitscan, or like a laser pointer?

it's a projectile but honestly i think i should make it a hitscan. making it laser-pointer-esque is also actually a really good idea, but i don't really want to rework everything so i'll probably just make it a hitscan when i address all this other stuff

**Comment by SimplySmiley123** [comment]
This is hilarious and unlocked a core memory in my dumb lizard brain about the old Splinter Cell games.

Definitely should be a black market item, doubt it would be ever bought if it was on the uplink just to kill lights one at a time.

**Comment by SomeoneYouProbablyKnow** [comment]
It could be put on the uplink but not as a standalone item. Maybe as part of the spy kit or some other kit? Or maybe as a free gift for buying thermals?

**Comment by HoodMilk** [comment]
Black market seems like a good place for this thing.

Does it leave a message in chat when you fire it?

**Comment by carlarctg** [comment]
You can add it smuggler satchels and contraband crates as well.

**Comment by ArcaneMusic** [comment]
God this is so based

**Comment by carlarctg** [comment]
I kinda want it to temporarily disable cameras as well, that'd be pretty neat and almost useful. But it risks making this a Balance Consideration.

**Comment by Ghommie** [comment]
> I kinda want it to temporarily disable cameras as well, that'd be pretty neat and almost useful. But it risks making this a Balance Consideration.

I don't think it'd an issue at all since it's a temporary effect.

**Review by Wallemations — APPROVED** [review]
They call it the black market because it's completely dark there from having no lightbulbs

**Comment by Hatterhat** [comment]
i'm gonna be real this is probably a disaster but it does all the things i want it to do

File: code/modules/projectiles/projectile/special/lightbreaker.dm [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+		if(attached?.light)
+			SEND_SIGNAL(attached.parent, COMSIG_HIT_BY_SABOTEUR, disrupt_duration)
+			lights_flickered++
+	if(lights_flickered)
+		to_chat(target, span_warning("Your light [lights_flickered > 1 ? "sources flick" : "source flicks"] off."))
+
+/obj/projectile/energy/fisher/melee
+	range = 1
+	suppressed = SUPPRESSED_VERY
+	disrupt_duration = 5 SECONDS
```
Comment (Ghommie) [inline code/modules/projectiles/projectile/special/lightbreaker.dm:41]: at least half a minute, I think.

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Review by Ghommie — APPROVED** [review]

**Comment by Ghommie** [comment]
The close range attack is a bit clunky, and makes quite a bit of noise and probably sends a message. I wonder if adding the NOBLUDGEON item flag to it would fix that. Beside, lgtm.

**Comment by Hatterhat** [comment]
someone suggested i make it do something to ethereals dont merge this until i figure out a good approach for that

**Comment by timothymtorres** [comment]
This may have caused:

- #78298

**Pull request merged**